### DISCUSSION TITLE
Lumped Mass Time Derivative Kernel with AD

### DISCUSSION THREAD

**Discussion by msbookwala** [idaholab/moose#28855] [Q&A General] [answered]
### Check these boxes if you have followed the posting rules.

- [X] Q&A General is the most appropriate section for my question
- [X] I have consulted the posting Guidelines on the Discussions front page
- [X] I have searched the Discussions forum and my question has not been asked before
- [X] I have searched the MOOSE website and the documentation does not answer my question
- [X] I have formatted my post following the posting guidelines (screenshots as a last resort, triple back quotes around pasted text)

### Question

I am trying to create a Lumped Mass Time Derivative Kernel with AD. For that I need $\dot{u}$ values on nodes. 
In MassLumpedTimeDerivative kernel, it has been defined using` _u_dot_nodal(_var.dofValuesDot())` in the constructor.
However,  with AD, only the methods `adDofValues()` and `adUDot()` are available, and not both. How can I define a Lumped mass Time derivative with AD? Is there any workaround?

**Comment by GiudGiud** [comment, marked as answer]
Hello

you're right this is not implemented
I just added this here if you want to try it out
https://github.com/idaholab/moose/pull/28857

I dont have a test for it yet though

Guillaume

**Reply by msbookwala** [reply]
Thanks a lot Guillaume. I'll try it and let you know.

Murtaza

**Reply by GiudGiud** [reply]
It should be good now. There was an issue if you pulled my branch yesterday

**Reply by msbookwala** [reply]
Hi, follow up on this issue. I implement a moose object ADMassLumpedTimeDerivative in the following way
Header file:
```c++
#pragma once

#include "ADTimeKernel.h"

class ADMassLumpedTimeDerivative : public ADTimeKernel
{
public:
  static InputParameters validParams();

  ADMassLumpedTimeDerivative(const InputParameters & parameters);

protected:
  virtual ADReal computeQpResidual() override;
  const ADVariableValue & _ad_u_dot_nodal;
  using ADKernel::_var;
};
``` 
Src file:
```c++

#include "ADMassLumpedTimeDerivative.h"

// MOOSE includes
#include "Assembly.h"
#include "MooseVariableFE.h"

#include "libmesh/quadrature.h"

registerMooseObject("discopanterApp", ADMassLumpedTimeDerivative);

InputParameters
ADMassLumpedTimeDerivative::validParams()
{
  InputParameters params = ADTimeKernel::validParams();
  params.addClassDescription(
      "Lumped AD formulation of the time derivative $\\frac{\\partial u}{\\partial t}$. Its "
      "corresponding weak form is $\\dot{u_i}(\\psi_i, 1)$ where $\\dot{u_i}$ denotes the time "
      "derivative of the solution coefficient associated with node $i$.");
  return params;
}

ADMassLumpedTimeDerivative::ADMassLumpedTimeDerivative(const InputParameters & parameters)
  : ADTimeKernel(parameters), _ad_u_dot_nodal(_var.adDofValuesDot())
{
}

ADReal
ADMassLumpedTimeDerivative::computeQpResidual()
{
  return _test[_i][_qp] * _ad_u_dot_nodal[_i];
}

``` 

However, I am getting the wrong Jacobian. What am I doing wrong here? 
I also get the following error when I am in debugger:

> Assertion `compare_dofs_set == current_dofs_set' failed
> We're going to see whether the dof sets are the same. IIRC the degree of freedom dependence (as indicated by the dof index set held by the ADReal) has to be the same for every residual passed to this method otherwise constrain_element_matrix will not work.

I can send the whole trace if needed.

**Reply by GiudGiud** [reply]
I ll take a look tomorrow. I verified we had the right dof dot values, I assumed the derivatives would be correct since they are pretty simple

**Reply by GiudGiud** [reply]
Actually looking at the object I just don't think you need AD for this. 

You can easily implement the Jacobian and use the non AD attribute

**Reply by msbookwala** [reply]
Hi.
We compared using TimeDerivative and ADTimeDerivative in our app and found a big difference in speed of codes, hence the motivation to use AD. However, we also need conservation of mass on the dofs. Hence the necessity of lumped mass formulation.

**Reply by GiudGiud** [reply]
You can implement lumpedMass without AD though.

**Reply by msbookwala** [reply]
Hi, 
There's already an implementation of a lumped mass time derivative. I just wanted to try lumped mass with AD to see if we get any speed gains for our app

**Reply by GiudGiud** [reply]
No AD is not going to be faster for this. AD helps with getting the Jacobian right, but hardcoding the Jacobian is actually better for speed.
